src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,125 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "random/generators.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+
+namespace cloud_storage {
+
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
+  : _cache_dir(std::move(cache_dir))
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period) {}
+
+ss::future<> cache::start() {
+    // todo: start maintenance cycle here
+    vlog(cst_log.debug, "max_cache_size: {}", _max_cache_size);
+    vlog(cst_log.debug, "check_period: {}", _check_period.count());
+    return ss::make_ready_future<>();
+}
+
+ss::future<> cache::stop() { return _gate.close(); }
+
+ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    gate_guard guard{_gate};
+    ss::file cache_file;
+    try {
+        cache_file = co_await ss::open_file_dma(
+          (_cache_dir / key).native(), ss::open_flags::ro);
+    } catch (...) {
+        vlog(
+          cst_log.debug,
+          "Couldn't open file {} to read: {}",
+          (_cache_dir / key).native(),
+          std::current_exception());
+        co_return std::nullopt;
+    }
+
+    auto data_size = co_await cache_file.size();
+    auto data_stream = ss::make_file_input_stream(cache_file);
+    co_return std::optional(cache_item{std::move(data_stream), data_size});
+}
+
+ss::future<>
+cache::put(std::filesystem::path key, ss::input_stream<char> data) {
+    gate_guard guard{_gate};
+
+    auto filename = (_cache_dir / key).filename();
+    auto tmp_filename = std::filesystem::path(
+      filename.string() + random_generators::gen_alphanum_string(4).data()
+      + ".part");
+    auto dir_path = (_cache_dir / key).remove_filename();
+
+    ss::output_stream<char> out;
+    try {
+        co_await ss::recursive_touch_directory(dir_path.string());
+
+        auto flags = ss::open_flags::rw | ss::open_flags::create
+                     | ss::open_flags::truncate;
+        auto cache_file = co_await ss::open_file_dma(
+          (dir_path / tmp_filename).native(), flags);
+        out = co_await ss::make_file_output_stream(cache_file);
+        co_await ss::copy(data, out);
+
+        co_await ss::rename_file(
+          (dir_path / tmp_filename).native(), (dir_path / filename).native());
+    } catch (...) {
+        vlog(
+          cst_log.error,
+          "Couldn't put data {} into cache: {}.",
+          (_cache_dir / key).native(),
+          std::current_exception());
+    }
+
+    co_await data.close();
+    co_await out.close();
```

Comment:
is it safe to close the output stream if it were never opened (e.g. based on what happens in try{})?

Reply:
i think this is relevant: https://github.com/vectorizedio/redpanda/pull/1906#discussion_r709229631 thanks @jcsp

Reply:
this might be relevant: https://github.com/vectorizedio/redpanda/pull/1906#discussion_r709229631 you could use perhaps `ss::defer` to make sure close is always called in the try block after the output stream is opened. i think the previous question about closing it if it hadn't been opened initially is also still standing.

Reply:
`close()` flushes and closes the underlying data source [doc](http://docs.seastar.io/master/classseastar_1_1input__stream.html#a3aff95b4e833ec2459ff276f9aa45b6f) in my case that's file  

IIUC closing stream that was not opened is as safe as closing file that was not opened.